### ISSUE TITLE
Levenberg Marquardt Learning Run Epoch Error Meaning

### ISSUE THREAD

**Issue by AlexGustafsson** [accord-net/framework#219]
I'm having a hard time wrapping my head around the meaning of the Error that is returned when i call
```
var teacher = new LevenbergMarquardtLearning(network);
var error = teacher.RunEpoch(inputs, outputs);
```

The error (for my data set anyways) go from 2000-4000 at the first epoch and drop down to 300-150 before it stops to change noteworthy.

I'm used to seeing error as a double value which is easy to correlate to percentages. So what does the error value actually mean?

**Comment by cesarsouza** [comment]
Hi Alex,

As of the current version, the Neuro learning algorithms report the sum-of-squared-errors as the error value returned by the RunEpoch functions. Basically, it amounts to computing the difference between the expected values and the network outputs, squaring, then summing across the dimensions of the output vector and across all samples in the dataset.

Please note that this has been carried over from the previous AForge.NET interfaces (a project from which Accord.NET was based upon). If you have suggestions for a more appropriate metric to be returned by those methods, please let me know.

Regards,
Cesar

**Comment by ChangePlaces** [comment]
I'm experimenting with neural networks at the moment, and after feeding in 1000+ training samples, the error rate has gone from 25600 to 4200, but the actual output has not changed, and still gives the exact same amount of correct answers and same amount of errors (59 correct, 1060 incorrect).

I'd have expected the amount of correct predictions to be somewhat reflected by the dramatic change in the error rate!

**Comment by cesarsouza** [comment]
Ooops, I wonder how I missed this update back from May, sorry about that!

@ChangePlaces this might be happening because the neural network error rate in the framework is always rate is always measured in terms of the sum-of-squared-errors. Under this circumstance, a smaller error rate does not necessarily mean that you will get a better classifier. A smaller sum-of-squared-error, in this case, simply indicates that the outputs of the network are getting closers to a 0 or 1 when making a binary decision, but note that actually this is not really necessary for making a correct decision. A network could make decisions which are just in the range of 0.4-0.6 (meaning that anything higher than 0.5 is a true and less is a false) and still be as good as a network that is doing decisions in the range (0-1).

Anyways, this behavior will change once Accord.Neuro is deprecated in favor of [Keras Sharp](https://github.com/cesarsouza/keras-sharp).
